File: IT/1/Oggetti/Esercizi/C++/esercizio25.cpp
/* file esercizio25.cpp */

#include <iostream>
#include <cmath>
#include <iomanip>

using std::cin;
using std::cout;
using std::endl;
using std::setprecision;
using std::ostream;

class Point2d {
	private:
		double x;
		double y;

	public:
		Point2d(double, double);
		double getX() const;
		void setX(double);
		double getY() const;
		void setY(double);

	friend ostream& operator<<(ostream& os, const Point2d&);
};

Point2d::Point2d(double x, double y) {
	this->setX(x);
	this->setY(y);
}

double Point2d::getX() const {
	return this->x;
}

void Point2d::setX(double x) {
	this->x = x;
}

double Point2d::getY() const {
	return this->y;
}

void Point2d::setY(double y) {
	this->y = y;
}

ostream& operator<<(ostream& os, const Point2d& p) {
	return os << "(" << p.x << ", " << p.y << ")";
}

class Point3d : public Point2d {
	private:
		double z;

	public:
		Point3d(double, double, double);
		double getZ();
		void setZ(double);

	friend ostream& operator<<(ostream& os, const Point3d&);
};

Point3d::Point3d(double x, double y, double z) : Point2d(x, y) {
	this->setZ(z);
}

double Point3d::getZ() {
	return this->z;
}

void Point3d::setZ(double z) {
	this->z = z;
}

ostream& operator<<(ostream& os, const Point3d& p) {
	return os << "(" << p.getX() << ", " << p.getY() << ", " << p.z << ")";
}

class Segment {
	private:
		Point3d* a;
		Point3d* b;

	public:
		Segment(Point3d*, Point3d*);
		Point3d* getA();
		void setA(Point3d*);
		Point3d* getB();
		void setB(Point3d*);
		double length();
};

Segment::Segment(Point3d* a, Point3d* b) {
	this->setA(a);
	this->setB(b);
}

Point3d* Segment::getA() {
	return this->a;
}

void Segment::setA(Point3d* a) {
	this->a = a;
}

Point3d* Segment::getB() {
	return this->b;
}

void Segment::setB(Point3d* b) {
	this->b = b;
}

double Segment::length() {
	return sqrt(pow((this->a->getX() - this->b->getX()), 2.0) + pow((this->a->getY() - this->b->getY()), 2.0) + pow((this->a->getZ() - this->b->getZ()), 2.0));
}

int main() {
	Point3d o(0.0, 0.0, 0.0);
	Point3d p(1.0, 2.0, 3.0);
	Segment s(&o, &p);

	cout << "O = " << o << " P = " << p << endl;
	cout << "Distanza tra i due punti: " << setprecision(3) << s.length() << " centimetri." << endl;
}

File: IT/1/Oggetti/Esercizi/Python/esercizio25.py
# file esercizio25.py

from __future__ import print_function
from math import pow, sqrt

class Point2d(object):
	def __init__(self, x, y):
		self.__x = x
		self.__y = y

	def getX(self):
		return self.__x

	def setX(self, x):
		self.__x = x

	def getY(self):
		return self.__y

	def setY(self, y):
		self.__y = y

	def __str__(self):
		return "(x={}, y={})".format(self.__x, self.__y)

	x = property(getX, setX)
	y = property(getY, setY)

class Point3d(Point2d):
	def __init__(self, x, y, z):
		super(Point3d, self).__init__(x, y)
		self.__z = z

	def getX(self):
		return self.__x

	def setZ(self, z):
		self.__z = z

	def getZ(self):
		return self.__z

	def __str__(self):
		return "(x={}, y={}, z={})".format(self.x, self.y, self.z)

	z = property(getZ, setZ)

class Segment(object):
	def __init__(self, a, b):
		self.__a = a
		self.__b = b

	def getA(self):
		return self.__a

	def setA(self, a):
		self.__a = a

	def getB(self):
		return self.__b

	def setB(self, b):
		self.__b = b

	def length(self):
		return sqrt((self.a.x - self.b.x)**2 + (self.a.y - self.b.y)**2 + (self.a.z - self.b.z)**2)

	a = property(getA, setA)
	b = property(getB, setB)

o = Point3d(0, 0, 0)
p = Point3d(1, 2, 3)
s = Segment(a=o, b=p)

print("O = {}, P = {}".format(o, p))
print("Distanza tra i due punti: {:.2f} centimetri.".format(s.length()))

